src/analyzer/plugins/analyzerkeyfinder.cpp

```diff
@@ -0,0 +1,123 @@
+#include "analyzer/plugins/analyzerkeyfinder.h"
+
+#include "analyzer/constants.h"
+#include "util/assert.h"
+#include "util/math.h"
+
+using mixxx::track::io::key::ChromaticKey;
+using mixxx::track::io::key::ChromaticKey_IsValid;
+
+namespace mixxx {
+
+AnalyzerKeyFinder::AnalyzerKeyFinder()
+        : m_pKeyFinder(nullptr),
+          m_pWorkspace(nullptr),
+          m_pAudioData(nullptr),
+          m_currentFrame(0),
+          m_previousKey(mixxx::track::io::key::INVALID) {
+}
+
+AnalyzerKeyFinder::~AnalyzerKeyFinder() {
+}
+
+bool AnalyzerKeyFinder::initialize(int samplerate) {
+    m_resultKeys.clear();
+    m_currentFrame = 0;
+    m_previousKey = mixxx::track::io::key::INVALID;
+
+    m_pKeyFinder = std::make_unique<KeyFinder::KeyFinder>();
+    m_pWorkspace = std::make_unique<KeyFinder::Workspace>();
+    m_pAudioData = std::make_unique<KeyFinder::AudioData>();
+    m_pAudioData->setFrameRate(samplerate);
+    m_pAudioData->setChannels(kAnalysisChannels);
+
+    return true;
+}
+
+bool AnalyzerKeyFinder::processSamples(const CSAMPLE* pIn, const int iLen) {
+    DEBUG_ASSERT(iLen % kAnalysisChannels == 0);
+    if (m_pAudioData->getSampleCount() == 0) {
+        m_pAudioData->addToSampleCount(iLen);
+    }
+
+    const size_t numInputFrames = iLen / kAnalysisChannels;
+    m_currentFrame += numInputFrames;
+
+    for (unsigned int frame = 0; frame < numInputFrames; frame++) {
+        for (unsigned int channel = 0; channel < kAnalysisChannels; channel++) {
```

Comment:
Could we use a single loop and div/mod by kAnalysisChannels instead of two nested loops?

Reply:
We could use a single loop simply looping over each sample one by one. We couldn't use a single loop referencing frame numbers without baking in an assumption about the number of channels (for example like [this](https://github.com/mixxxdj/mixxx/blob/master/src/effects/builtin/bitcrushereffect.cpp#L106)). I think it would be a good idea to move towards this nested loop pattern of iterating over frames and channels throughout Mixxx.